src/reader.cpp

```diff
@@ -1136,17 +1136,11 @@ static history_pager_result_t history_pager_search(const HistorySharedPtr &histo
     size_t page_size = std::max(termsize_last().height / 2 - 2, (rust::isize)12);
 
     rust::Box<completion_list_t> completions = new_completion_list();
+
     rust::Box<HistorySearch> search =
-        rust_history_search_new(history, search_string.c_str(), history_search_type_t::Contains,
+        rust_history_search_new(history, search_string.c_str(), history_search_type_t::ContainsGlob,
```

Comment:
if the value is from the commandline, we need to escape it:

```
diff --git a/src/reader.cpp b/src/reader.cpp
index b10700988..6c45d6123 100644
--- a/src/reader.cpp
+++ b/src/reader.cpp
@@ -3672,7 +3672,8 @@ void reader_data_t::handle_readline_command(readline_cmd_t c, readline_loop_stat
             pager.set_prefix(MB_CUR_MAX > 1 ? L"► " : L"> ", false /* highlight */);
             // Update the search field, which triggers the actual history search.
             if (!history_search.active() || history_search.search_string().empty()) {
-                insert_string(pager.search_field_line(), *command_line.text());
+                insert_string(pager.search_field_line(),
+                              parse_util_escape_wildcards(*command_line.text()));
             } else {
                 // If we have an actual history search already going, reuse that term
                 // - this is if the user looks around a bit and decides to switch to the pager.
```

(need to add `parse_util_escape_wildcards`)

Reply:
> Question 1: If the user decide to cancel the history pager (by pressing enter on no match), this will result in the user getting the escaped chars on their command line. So we need to unescape the wildcards again, right?

Good point, yes we should probably unescape the search text before copying it to the commandline. Not a big deal though. I'm not sure what's the least surprising behavior overall.

> Or is it possible to get the behavior as if user pressed esc so that their original input is restored?

Escape (aka `rl::cancel`) uses undo to restore the original input, but on Enter we want to copy whatever is in the pager search field at the moment, not necessarily the original commandline.

> Question 2: the semantic of the word "escape" in the parse_util is not very consistent.
> IMHO, what the current parse_util_unescape_wildcards does is not what I would call "unescape", at least not as in that escape_wildcards and unescape_wildcards are the inverse function of the other one.
> Perhaps we should name parse_util_escape_wildcards differently?

Yeah `parse_util_escape_wildcards` works on special internal representation of globs. 
We could put the one with the behavior we want here under `escape_string`.

LGTM overall.

Reply:
Thanks. Is there anything else I should do in this PR before it was accepted? I can add a note about the change in `CHANGELOG.rst`, but I'm not sure under which version I put it. I saw this PR is tagged [fish next-3.x](https://github.com/fish-shell/fish-shell/milestone/36), but I wasn't so sure about what that means.

Reply:
Add it under the "fish 3.8.0" section, I've just cleaned it up to allow that.